zipline/pipeline/hooks/progress.py

```diff
@@ -0,0 +1,494 @@
+"""Pipeline hooks for tracking and displaying progress.
+"""
+import cgi
+from collections import namedtuple
+import time
+
+from interface import implements
+
+from zipline.utils.compat import contextmanager
+from zipline.utils.string_formatting import bulleted_list
+
+from .iface import PipelineHooks
+
+
+class ProgressHooks(implements(PipelineHooks)):
+    """
+    Hooks implementation for displaying progress.
+
+    Parameters
+    ----------
+    publisher_factory : callable
+        Function producing a new object with a ``publish()`` method that takes
+        a ``ProgressModel`` and publishes progress to a consumer.
+    """
+    def __init__(self, publisher_factory):
+        # This object encapsulates state that should be reset between pipeline
+        # executions.
+        self._state = _ProgressHooksState(publisher_factory)
```

Comment:
I think my idea here was to encapsulate all the state that needed to be reset between each iteration, so that the reset process could just be "create a new object and set one attribute" rather than "reset a whole bunch of attributes", which always makes me slightly paranoid about failing to reset properly and having state leak between iterations.

At some point in my earlier drafts I switched to having the state have a `reset` method though...I'm not sure why. I've inlined this for now. If the resetting logic here gets more complex, I think there is potentially value in encapsulating the transient state more thoroughly.